src/linux_native.rs

```diff
@@ -0,0 +1,670 @@
+        match key_cmd {
+            // Usage Page 6.2.2.7 (Global)
+            0x4 => {
+                usage_page = hid_report_bytes(desc, data_len) as u16;
+            }
+            // Usage 6.2.2.8 (Local)
+            0x8 => {
+                usage = Some(hid_report_bytes(desc, data_len) as u16);
+            }
+            // Collection 6.2.2.4 (Main)
+            0xa0 => {
+                // Usage is a Local Item, unset it
+                if let Some(u) = usage.take() {
+                    usage_pair = Some((usage_page, u))
+                }
+            }
+            // Input 6.2.2.4 (Main)
+		        0x80 |
+            // Output 6.2.2.4 (Main)
+		        0x90 |
+            // Feature 6.2.2.4 (Main)
+		        0xb0 |
+            // End Collection 6.2.2.4 (Main)
+		        0xc0  =>  {
+			          // Usage is a Local Item, unset it
+                usage.take();
+            }
+            _ => {}
+        }
+
+        advanced += data_len + key_size;
+        desc = &desc[(data_len + key_size)..];
+        if let Some((usage_page, usage)) = usage_pair {
+            return Some((usage_page, usage, advanced));
+        }
+    }
+
+    if let (true, Some(usage)) = (initial, usage) {
+        return Some((usage_page, usage, advanced));
+    }
+
+    None
+}
+
+/// Gets the size of the HID item at the given position
+///
+/// Returns data_len and key_size when successful
+fn hid_item_size(desc: &[u8]) -> Option<(usize, usize)> {
+    let key = desc[0];
+
+    // Long Item. Next byte contains the length of the data section.
+    if (key & 0xf0) == 0xf0 {
+        if !desc.is_empty() {
+            return Some((desc[1].into(), 3));
+        }
+
+        // Malformed report
+        return None;
+    }
+
+    // Short Item. Bottom two bits contains the size code
+    match key & 0x03 {
+        v @ 0 | v @ 1 | v @ 2 => Some((v.into(), 1)),
+        3 => Some((4, 1)),
+        _ => unreachable!(), // & 0x03 means this can't happen
+    }
+}
+
+/// Get the bytes from a HID report descriptor
+///
+/// Must only be called with `num_bytes` 0, 1, 2 or 4.
+fn hid_report_bytes(desc: &[u8], num_bytes: usize) -> u32 {
+    if num_bytes >= desc.len() {
+        return 0;
+    }
+
+    match num_bytes {
+        0 => 0,
+        1 => desc[1] as u32,
+        2 => {
+            let bytes = [desc[1], desc[2], 0, 0];
+            u32::from_le_bytes(bytes)
+        }
+        4 => u32::from_le_bytes(desc[1..=4].try_into().unwrap()),
+        _ => unreachable!(),
+    }
+}
+
+/// Get the attribute from the device and convert it into a [`WcharString`].
+fn attribute_as_wchar(dev: &udev::Device, attr: &str) -> WcharString {
+    dev.attribute_value(attr)
+        .map(Into::into)
+        .map(osstring_to_string)
+        .unwrap_or(WcharString::None)
+}
+
+/// Get the attribute from the device and convert it into a i32
+///
+/// On error or if the attribute is not found, it returns -1;
+fn attribute_as_i32(dev: &udev::Device, attr: &str) -> i32 {
```

Comment:
I would rather have API`s using Option / Result instead of negative error codes.

Reply:
Yeah, that'd make sense as a rustification pass.